src/extract_gpuinfo_msm.c

```diff
@@ -0,0 +1,523 @@
+/*
+ *
+ * Copyright (C) 2023 Ryan Houdek <Sonicadvance1@gmail.com>
+ *
+ * This file is part of Nvtop and adapted from the amdgpu implementation.
+ *
+ * Nvtop is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Nvtop is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with nvtop.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "nvtop/device_discovery.h"
+#include "nvtop/extract_gpuinfo_common.h"
+#include "nvtop/extract_processinfo_fdinfo.h"
+#include "nvtop/time.h"
+
+#include <assert.h>
+#include <dlfcn.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <libdrm/msm_drm.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/sysinfo.h>
+#include <unistd.h>
+#include <uthash.h>
+#include <xf86drm.h>
+
+// extern
+const char * msm_parse_marketing_name(uint64_t gpu_id);
+
+#define HASH_FIND_CLIENT(head, key_ptr, out_ptr) HASH_FIND(hh, head, key_ptr, sizeof(unsigned), out_ptr)
+#define HASH_ADD_CLIENT(head, in_ptr) HASH_ADD(hh, head, client_id, sizeof(unsigned), in_ptr)
+
+#define SET_MSM_CACHE(cachePtr, field, value) SET_VALUE(cachePtr, field, value, msm_cache_)
+#define RESET_MSM_CACHE(cachePtr, field) INVALIDATE_VALUE(cachePtr, field, msm_cache_)
+#define MSM_CACHE_FIELD_VALID(cachePtr, field) VALUE_IS_VALID(cachePtr, field, msm_cache_)
+
+enum intel_process_info_cache_valid {
+  msm_cache_engine_render_valid = 0,
+  msm_cache_process_info_cache_valid_count
+};
+
+struct __attribute__((__packed__)) unique_cache_id {
+  unsigned client_id;
```

Comment:
Can there be multiple Adreno GPUs on one system? If yes, is the client-id global?
[Because the doc either propose the client-id to be unique or having the client-id+pdev tuple be unique](https://docs.kernel.org/6.1/gpu/drm-usage-stats.html#optional-fully-standardised-keys)

Reply:
So far only a single one, and I wouldn't expect that to change unless qcom decides to start making dgpu's..  if there is no pdev field then the client-id is globally unique.

(In practice, my patchset that adds GPU mem info also makes client-id globally unique in all cases, because that is a simpler implementation on the kernel side.)